tests/e2e/utils/data_fetcher/traces.py

```diff
@@ -0,0 +1,266 @@
+from botocore.paginate import PageIterator
+from mypy_boto3_xray.client import XRayClient
+from mypy_boto3_xray.type_defs import TraceSummaryTypeDef
+from pydantic import BaseModel
+from retry import retry
+
+
+class TraceSubsegment(BaseModel):
+    id: str  # noqa: A003 VNE003  # id is a field we can't change
+    name: str
+    start_time: float
+    end_time: float
+    aws: Optional[dict]
+    subsegments: Optional[List["TraceSubsegment"]]
+    annotations: Optional[Dict[str, Any]]
+    metadata: Optional[Dict[str, Dict[str, Any]]]
+
+
+class TraceDocument(BaseModel):
+    id: str  # noqa: A003 VNE003  # id is a field we can't change
+    name: str
+    start_time: float
+    end_time: float
+    trace_id: str
+    parent_id: Optional[str]
+    aws: Dict
+    origin: str
+    subsegments: Optional[List[TraceSubsegment]]
+
+
+class TraceFetcher:
+    default_exclude_seg_name: List = ["Initialization", "Invocation", "Overhead"]
+
+    def __init__(
+        self,
+        filter_expression: str,
+        start_date: datetime,
+        end_date: Optional[datetime] = None,
+        xray_client: Optional[XRayClient] = None,
+        exclude_segment_name: Optional[List[str]] = None,
+        resource_name: Optional[List[str]] = None,
+        origin: Optional[List[str]] = None,
+        minimum_traces: int = 1,
+    ):
+        """Fetch and expose traces from X-Ray based on parameters
+
+        Data is recursively fetched in the following order:
+
+        * Trace summaries
+        * Trace IDs
+        * Traces
+        * Segments
+        * Subsegments
+        * Nested Subsegments
+
+        Parameters
+        ----------
+        filter_expression : str
+            AWS X-Ray Filter Expressions
+            see: https://docs.aws.amazon.com/xray/latest/devguide/xray-console-filters.html
+        start_date : datetime
+            Start date range to filter traces
+        end_date : Optional[datetime], optional
+            End date range to filter traces, by default 5 minutes past start_date
+        xray_client : Optional[XRayClient], optional
+            AWS X-Ray SDK Client, by default boto3.client('xray')
+        exclude_segment_name : Optional[List[str]], optional
+            Name of segments to exclude, by default ["Initialization", "Invocation", "Overhead"]
+        resource_name : Optional[List[str]], optional
+            Name of resource to filter traces (e.g., function name), by default None
+        origin : Optional[List[str]], optional
+            Trace origin name to filter traces, by default ["AWS::Lambda::Function"]
+        minimum_traces : int
+            Minimum number of traces to be retrieved before exhausting retry attempts
+        """
+        self.filter_expression = filter_expression
+        self.start_date = start_date
+        self.end_date = end_date or self.start_date + timedelta(minutes=5)
+        self.xray_client: XRayClient = xray_client or boto3.client("xray")
+        self.trace_ids: List[str] = []
```

Comment:
Is it used here? I can see that we overwrite trace_ids, trace_documents and subsegments in lines 104-106